INET data type backward compatibility problem when specifying a network address

This bug report is based on recent changes to driver version 42.2.7 based on the following PR dicsussions and merged changes: https://github.com/pgjdbc/pgjdbc/issues/1134 https://github.com/pgjdbc/pgjdbc/pull/1527

  <!--- What kind of an issue is this? Put an `x` in all the boxes that apply: -->
- [x] bug report
- [ ] feature request
    

**Describe the issue**
```
Caused by: org.postgresql.util.PSQLException: IP address 192.168.0.0/16 of a host could not be determined
    at org.postgresql.jdbc.PgConnection.getObject(PgConnection.java:558)
    at org.postgresql.jdbc.PgResultSet.getInetAddress(PgResultSet.java:3101)
    at org.postgresql.jdbc.PgResultSet.internalGetObject(PgResultSet.java:221)
    at org.postgresql.jdbc.PgResultSet.getObject(PgResultSet.java:2566)
    ... 8 more
Caused by: java.net.UnknownHostException: 192.168.0.0/16
    at java.base/java.net.InetAddress$CachedAddresses.get(InetAddress.java:792)
    at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1494)
    at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1353)
    at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1287)
    at java.base/java.net.InetAddress.getByName(InetAddress.java:1237)
    at org.postgresql.jdbc.PgConnection.getObject(PgConnection.java:556)
    ... 14 more
```
A table is defined with an `INET` data type and an insert is being done into said table with a network-based value of `192.168.0.0/16` leads to a `PSQLException` when querying table. 

**Driver Version?** 
42.2.7

Validated this was not an issue again 42.2.5 and 42.2.6 driver versions.

**Java Version?**
9

**OS Version?**
Fedora 30

**PostgreSQL Version?**
10

**To Reproduce**
Steps to reproduce the behaviour:
1. Create table with `INET` data type
2. Insert record with a network-based address `192.168.0.0/16`
3. Run a query against the table and get exception shown above

**Expected behaviour**
The `inet` value would have been returned without exception.

--- a/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java
+++ b/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java
@@ -39,8 +39,6 @@ import org.postgresql.util.PSQLException;
 import org.postgresql.util.PSQLState;
 
 import java.io.IOException;
-import java.net.InetAddress;
-import java.net.UnknownHostException;
 import java.sql.Array;
 import java.sql.Blob;
 import java.sql.CallableStatement;
@@ -548,15 +546,6 @@ public class PgConnection implements BaseConnection {
         // Handle the type (requires SQLInput & SQLOutput classes to be implemented)
         throw new PSQLException(GT.tr("Custom type maps are not supported."),
             PSQLState.NOT_IMPLEMENTED);
-      }
-    }
-
-    if (type.equals("inet")) {
-      try {
-        return InetAddress.getByName(value);
-      } catch (UnknownHostException e) {
-        throw new PSQLException(GT.tr("IP address {0} of a host could not be determined", value),
-          PSQLState.CONNECTION_FAILURE, e);
       }
     }
 

--- a/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java
+++ b/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java
@@ -36,6 +36,7 @@ import java.io.UnsupportedEncodingException;
 import java.math.BigDecimal;
 import java.math.BigInteger;
 import java.net.InetAddress;
+import java.net.UnknownHostException;
 import java.sql.Array;
 import java.sql.Blob;
 import java.sql.Clob;
@@ -215,10 +216,6 @@ public class PgResultSet implements ResultSet, org.postgresql.PGRefCursorResultS
         // if the backend doesn't know the type then coerce to String
         if (type.equals("unknown")) {
           return getString(columnIndex);
-        }
-
-        if (type.equals("inet")) {
-          return getInetAddress(getPGType(columnIndex), getString(columnIndex));
         }
 
         if (type.equals("uuid")) {
@@ -3095,16 +3092,6 @@ public class PgResultSet implements ResultSet, org.postgresql.PGRefCursorResultS
     return new UUID(ByteConverter.int8(data, 0), ByteConverter.int8(data, 8));
   }
 
-  protected Object getInetAddress(String type, String data) throws SQLException {
-    InetAddress inet;
-    try {
-      inet = (InetAddress) connection.getObject(type, data, null);
-    } catch (IllegalArgumentException iae) {
-      throw new PSQLException(GT.tr("Invalid Inet data."), PSQLState.INVALID_PARAMETER_VALUE, iae);
-    }
-    return inet;
-  }
-
   private class PrimaryKey {
     int index; // where in the result set is this primaryKey
     String name; // what is the columnName of this primary Key
@@ -3355,11 +3342,16 @@ public class PgResultSet implements ResultSet, org.postgresql.PGRefCursorResultS
     } else if (type == UUID.class) {
       return type.cast(getObject(columnIndex));
     } else if (type == InetAddress.class) {
-      Object addressString = getObject(columnIndex);
-      if (addressString == null) {
+      String inetText = getString(columnIndex);
+      if (inetText == null) {
         return null;
       }
-      return type.cast(getObject(columnIndex));
+      int slash = inetText.indexOf("/");
+      try {
+        return type.cast(InetAddress.getByName(slash < 0 ? inetText : inetText.substring(0, slash)));
+      } catch (UnknownHostException ex) {
+        throw new PSQLException(GT.tr("Invalid Inet data."), PSQLState.INVALID_PARAMETER_VALUE, ex);
+      }
       // JSR-310 support
       //#if mvn.project.property.postgresql.jdbc.spec >= "JDBC4.2"
     } else if (type == LocalDate.class) {

--- a/pgjdbc/src/test/java/org/postgresql/test/jdbc4/jdbc41/GetObjectTest.java
+++ b/pgjdbc/src/test/java/org/postgresql/test/jdbc4/jdbc41/GetObjectTest.java
@@ -22,6 +22,7 @@ import org.postgresql.geometric.PGpolygon;
 import org.postgresql.test.TestUtil;
 import org.postgresql.util.PGInterval;
 import org.postgresql.util.PGmoney;
+import org.postgresql.util.PGobject;
 
 import org.junit.After;
 import org.junit.Before;
@@ -938,42 +939,48 @@ public class GetObjectTest {
     }
   }
 
-  /**
-   * Test the behavior getObject for inet columns.
-   */
-  @Test
-  public void testGetInet4Address() throws SQLException, UnknownHostException {
+  private void testInet(String inet, InetAddress expectedAddr, String expectedText) throws SQLException, UnknownHostException {
+    PGobject expectedObj = new PGobject();
+    expectedObj.setType("inet");
+    expectedObj.setValue(expectedText);
     Statement stmt = conn.createStatement();
-    String expected = "192.168.100.128";
-    stmt.executeUpdate(TestUtil.insertSQL("table1","inet_column","'" + expected + "'"));
-
-    ResultSet rs = stmt.executeQuery(TestUtil.selectSQL("table1", "inet_column"));
+    ResultSet rs = stmt.executeQuery("SELECT '" + inet + "'::inet AS inet_column");
     try {
       assertTrue(rs.next());
-      assertEquals(InetAddress.getByName(expected), rs.getObject("inet_column", InetAddress.class));
-      assertEquals(InetAddress.getByName(expected), rs.getObject(1, InetAddress.class));
+      assertEquals("The string value of the inet should match when fetched via getString(...)", expectedText, rs.getString(1));
+      assertEquals("The string value of the inet should match when fetched via getString(...)", expectedText, rs.getString("inet_column"));
+      assertEquals("The object value of the inet should match when fetched via getObject(...)", expectedObj, rs.getObject(1));
+      assertEquals("The object value of the inet should match when fetched via getObject(...)", expectedObj, rs.getObject("inet_column"));
+      assertEquals("The InetAddress value should match when fetched via getObject(..., InetAddress.class)", expectedAddr, rs.getObject("inet_column", InetAddress.class));
+      assertEquals("The InetAddress value should match when fetched via getObject(..., InetAddress.class)", expectedAddr, rs.getObject(1, InetAddress.class));
     } finally {
       rs.close();
+      stmt.close();
     }
   }
 
   /**
-   * Test the behavior getObject for inet columns.
+   * Test the behavior getObject for ipv4 inet columns.
+   */
+  @Test
+  public void testGetInet4Address() throws SQLException, UnknownHostException {
+    String inet = "192.168.100.128";
+    InetAddress addr = InetAddress.getByName(inet);
+    testInet(inet, addr, inet);
+    testInet(inet + "/16", addr, inet + "/16");
+    testInet(inet + "/32", addr, inet);
+  }
+
+  /**
+   * Test the behavior getObject for ipv6 inet columns.
    */
   @Test
   public void testGetInet6Address() throws SQLException, UnknownHostException {
-    Statement stmt = conn.createStatement();
-    String expected = "2001:4f8:3:ba:2e0:81ff:fe22:d1f1";
-    stmt.executeUpdate(TestUtil.insertSQL("table1","inet_column","'" + expected + "'"));
-
-    ResultSet rs = stmt.executeQuery(TestUtil.selectSQL("table1", "inet_column"));
-    try {
-      assertTrue(rs.next());
-      assertEquals(InetAddress.getByName(expected), rs.getObject("inet_column", InetAddress.class));
-      assertEquals(InetAddress.getByName(expected), rs.getObject(1, InetAddress.class));
-    } finally {
-      rs.close();
-    }
+    String inet = "2001:4f8:3:ba:2e0:81ff:fe22:d1f1";
+    InetAddress addr = InetAddress.getByName(inet);
+    testInet(inet, addr, inet);
+    testInet(inet + "/16", addr, inet + "/16");
+    testInet(inet + "/128", addr, inet);
   }
 
 }

